packages/catalog-api/src/lib/schema.graphql

```diff
@@ -0,0 +1,237 @@
+  will just be defined by the database implementation.
+  """
+  elements(tag: String = "latest", limit: Int): [CustomElement!]!
+
+  element(packageName: String!, elementName: String!, tag: String = "latest"): CustomElement
+
+  """
+  Validates a package, returning a list of problems
+  """
+  validate(packageName: String!, tag: String = "latest"): [String!]!
+}
+
+type Mutation {
+  importPackage(packageName: String!): PackageInfo
+
+  deletePackage(packageName: String!): Boolean
+}
+
+"""
+Information about an npm package that's either version-independent
+or related to the latest version of the package.
+"""
+type PackageInfo {
+  name: ID!
+
+  status: PackageStatus!
+
+  lastUpdate: Date!
+
+  """
+  The package description derived from the latest version
+  """
+  description: String
+
+  """
+  The dist tag mappings for the package.
+  """
+  distTags: [DistTag!]! @map
+
+  """
+  Retreives a published versioned package. The default is "latest" for the
+  current version.
+  """
+  version(versionOrTag: String = "latest"): PackageVersion @collection
+}
+
+type DistTag {
+  tag: String!
+  version: String!
+}
+
+"""
+The status of a package.
+
+Concurrent requests for package metadata, which may trigger fetching from
+npm, are possible. We need to ensure there's only one writing request via
+transactions that lock on the status field.
+
+Because packages have versions that update over a time, a package may have
+a pending mutation task from initial fetching, or updating by querying and
+fetching new version metadata.
+
+Packages may not exist, in which case we record that they're not found in
+order to avoid flooding npm with invalid requests.
+
+Errors while fetching packages are divided into two categories:
+ - Invalid packages for which we were able to fetch metadata, but something
+in it was incorrect. Hopefully this never happens, but is included for
+completeness.
+ - Errors, such as network failures. In these cases we should be able to
+retry fetching package metadata.
+
+TODO: we need time and possibly version associtated with the last impmort
+attempt.
+"""
+enum PackageStatus {
+  """
+  The package is being downloaded and indexed for the first time.
+  """
+  INITIALIZING
+
+  """
+  Packages may be read from, but there is a pending update task and
+  new versions of the package are being downloaded and indexed.
+  """
+  UPDATING
+
+  """
+  The package is indexed and read to be read.
+  """
+  READY
+
+  """
+  The package was not found on npm.
+  """
+  NOT_FOUND
+
+  """
+  The package metadata was invalid. This is not a recoverable error unless a
+  new package version is published.
+  """
+  INVALID
+
+  """
+  An other recoverable error occured, such as network failure.
+  """
+  ERROR
+}
+
+"""
+Information about a specific version of a package
+"""
+type PackageVersion {
+  version: String!
+
+  """
+  The dist tags assigned to this version
+  """
+  distTags: [String!]!
```

Comment:
what happens when we notice that a distTag is moved to a new version? we write to both PackageVersions?

Reply:
Yeah. This is what I'm working on right now. We only know the dist-tags when we import a _package_, and then because we denormalize we have to update all _package versions_ and all custom elements that had a dist-tag change. Since we store the dist-tags, we can calculate if/when we need to do that.